doc/mgmt/SONiC_YANG_Model_Guidelines.md

```diff
@@ -0,0 +1,961 @@
+
+#   SONiC YANG MODEL GUIDELINES
+
+
+## Revision
+
+ | Rev |     Date    |       Author       | Change Description                |
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 1.0 | 22 Aug 2019 | Praveen Chaudhary  | Initial version                   |
+ | 1.0 | 11 Sep 2019 | Partha Dutta       | Adding additional steps for SONiC YANG  |
+
+## References
+| References 		    |     Date/Version    |   		    Link     	 	|
+|:-------------------------:|:-------------------:|:-----------------------------------:|
+| RFC 7950                  | August 2016         | https://tools.ietf.org/html/rfc7950 |
+| Management Framework      | 0.9                 | TBD                                 |
+
+## Terminology and Acronyms
+| Acronyms 		    |    Description/Expansion      |
+|:-------------------------:|:-------------------:|
+| ABNF           | Augmented Backus-Naur Form     | 
+| XPath          | XML Path Language              |
+| CVL		 | Configuration Validation Library |
+
+
+## Overview
+ This document lists the guidelines, which will be used to write YANG Modules for SONiC. These YANG Modules (called SONiC YANG models) will be primarily based on or represent the ABNF.json of SONiC, and the syntax of YANG models must follow RFC 7950 ([https://tools.ietf.org/html/rfc7950](https://tools.ietf.org/html/rfc7950)).  Details of config in format of ABNF.json can be found at https://github.com/Azure/SONiC/wiki/Configuration.
+
+These YANG models will be used to verify the configuration for SONiC switches, so a library which supports validation of configuration on SONiC Switch must use these YANG Models. List of such Libraries are: 1.) Configuration Validation Library. (CVL).  YANG models, which are written using these guidelines can also be used as User End YANG Models, i.e North Bound configuration tools or CLI can provide config data in sync with these YANG models. 
+
+[SONiC Management Framework](https://github.com/Azure/SONiC/pull/436) uses SONiC YANG models for configuration management in two ways :
+
+* As *custom Northbound Management YANG* in absence of standard YANG model and for configuration validation (CVL uses it).
+* As *CVL YANG* for configuration validation only, when standard YANG model (Openconfig, IETF etc.) is used as Northbound Management YANG. In such case, steps below which are marked as "This step applies when SONiC YANG is used as Northbound YANG." can be skipped.
+
+
+## Guidelines
+
+
+### 1. All schema definitions related to a feature should be written in a single YANG model file. YANG model file is named as 'sonic-{feature}.yang', e.g. for ACL feature sonic-acl.yang is the file name. It is mandatory to define a top level YANG container named as 'sonic-{feature}' i.e. same as YANG model name. For example, sonic-acl.yang should have 'sonic-acl' as top level container. All other definition should be written inside the top level container.
```

Comment:
I feel, we should either add all new points at top of the document or at the last so that it is easy to track, what is being modified and what is added new.

Reply:
Please look into the diff to find out the changes.